### PR TITLE
network: route - add support to configure tcp advmss

### PR DESCRIPTION
**Pull request by ssahani**
closes https://github.com/systemd/systemd/issues/18055

### CODE REVIEW

File: man/systemd.network.xml
Comment (poettering) [inline man/systemd.network.xml:1476]: this is a size in bytes (network people would probably say "octet", but anyone else just calls it a byte), hence should say so.

File: src/network/networkd-route.c
Comment (poettering) [inline src/network/networkd-route.c]: should use parse_size(), similar to parse_mtu().

File: man/systemd.network.xml
Comment (poettering) [reply man/systemd.network.xml:1476]: hmm, the MSS is not the same as the MTU iirc, as the former is without IP headers. see https://en.wikipedia.org/wiki/Maximum_segment_size

hence maybe not claim that this was the path mtu

Comment (poettering) [reply man/systemd.network.xml:1476]: the size 0 is not allowed, is it?

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/networkd-route.c
Comment (poettering) [reply src/network/networkd-route.c]: (and then range checks enforced, and I think == 0 at least is not actually allowed, probably other limits apply too)

File: man/systemd.network.xml
Comment (ssahani) [reply man/systemd.network.xml:1476]: Yes it's in bytes. I updated the man.

**Comment by ssahani** [comment]
Updated thanks for the review!

**Comment by bluca** [comment]
> Updated thanks for the review!

I don't see the changes to the manpage?

**Comment by ssahani** [comment]
> > Updated thanks for the review!
> 
> I don't see the changes to the manpage?


I just added ```(in bytes)``` and k m g etc. https://github.com/systemd/systemd/pull/18131/commits/63a75cf83c29a918193b70ff7e4c71b5d5233712#diff-d81a981ca75657b8fc8f6ad6871cb9f8d7705944919b67dad745680ebeec1b4eR1474

Not sure why it's not showing. on the main page

**Comment by bluca** [comment]
> > > Updated thanks for the review!
> > 
> > 
> > I don't see the changes to the manpage?
> 
> I just added `(in bytes)` and k m g etc. [63a75cf#diff-d81a981ca75657b8fc8f6ad6871cb9f8d7705944919b67dad745680ebeec1b4eR1474](https://github.com/systemd/systemd/commit/63a75cf83c29a918193b70ff7e4c71b5d5233712#diff-d81a981ca75657b8fc8f6ad6871cb9f8d7705944919b67dad745680ebeec1b4eR1474)
> 
> Not sure why it's not showing. on the main page

It's there now, probably some gui issues.

The manpage still says MTU though, it should be changed to MSS as Lennart mentioned.

**Comment by ssahani** [comment]
done thanks !

File: man/systemd.network.xml
Comment (bluca) [reply man/systemd.network.xml:1476]: > the size 0 is not allowed, is it?

Size 0 is not rejected, it just means the value is not used. That seems fine to me. It's a way for users to reset it back I guess. Should it be rejeced as invalid instead?

Comment (ssahani) [reply man/systemd.network.xml:1476]: TCP segment size can't be zero . We don't actually check for 0 explicitly.  Just not used and will not be applied.

Comment (bluca) [reply man/systemd.network.xml:1476]: That sounds fine to me. Setting green label while waiting for CI, but will wait to see if @poettering is OK with this handling of the zero value before merging.

File: src/network/networkd-route.c
Comment (poettering) [inline src/network/networkd-route.c:2127]: so, we should filter out "0" here, and refuse it.

Our usual way how to unset an option is by assigning the empty string to it. Hence, please:

1. refuse TCPAdvertisedMaximumSegmentSize=0 here
2. support TCPAdvertisedMaximumSegmentSize= and in that case set the field to 0

i.e. implicit behaviour of some magic undocumented special values is not OK. We should make this explicit, i.e. empty string should universally mean "unset" or "reset to defaults". And if you specify a numeric value, then the it should be taken literally, and if that makes no sense (like in this case afaics if zero) be refused.

File: src/network/networkd-route.c
Comment (poettering) [inline src/network/networkd-route.c]: your are propagating the error here, not ignoring it.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/networkd-route.c
Comment (ssahani) [reply src/network/networkd-route.c:2127]: OK Updated. ptal. thanks .

**Comment by bluca** [comment]
Comments addressed, so flipping the labels again. Waiting for the reviewer to have a chance to look again.

**Comment by poettering** [comment]
lgtm

**Pull request merged**
